stdlib/public/core/StringInterpolation.swift

```diff
@@ -26,48 +127,133 @@ extension String : _ExpressibleByStringInterpolation {
+  /// string literal being created.
+  /// 
+  /// Do not call this method directly. It is used by the compiler when
+  /// interpreting string interpolations. Instead, use string
+  /// interpolation to create a new string by including values, literals,
+  /// variables, or expressions enclosed in parentheses, prefixed by a
+  /// backslash (`\(`...`)`).
   ///
-  /// Do not call this initializer directly. It is used by the compiler when
-  /// interpreting string interpolations.
+  ///     let price = 2
+  ///     let number = 3
+  ///     let message = """
+  ///                   If one cookie costs \(price) dollars, \
+  ///                   \(number) cookies cost \(price * number) dollars.
+  ///                   """
+  ///     print(message)
+  ///     // Prints "If one cookie costs 2 dollars, 3 cookies cost 6 dollars."
   @inlinable
-  public init<T>(stringInterpolationSegment expr: T) {
-    self = String(describing: expr)
+  public mutating func appendInterpolation<T>(_ value: T)
+    where T: CustomStringConvertible
+  {
+    value.description.write(to: &self)
   }
-
-  /// Creates a string containing the given value's textual representation.
+  
+  /// Interpolates the given value's textual representation into the
+  /// string literal being created.
+  /// 
+  /// Do not call this method directly. It is used by the compiler when
+  /// interpreting string interpolations. Instead, use string
+  /// interpolation to create a new string by including values, literals,
+  /// variables, or expressions enclosed in parentheses, prefixed by a
+  /// backslash (`\(`...`)`).
   ///
-  /// Do not call this initializer directly. It is used by the compiler when
-  /// interpreting string interpolations.
+  ///     let price = 2
+  ///     let number = 3
+  ///     let message = """
+  ///                   If one cookie costs \(price) dollars, \
+  ///                   \(number) cookies cost \(price * number) dollars.
+  ///                   """
+  ///     print(message)
+  ///     // Prints "If one cookie costs 2 dollars, 3 cookies cost 6 dollars."
   @inlinable
-  public init<T: TextOutputStreamable> (stringInterpolationSegment expr: T) {
-    self = _toStringReadOnlyStreamable(expr)
+  public mutating func appendInterpolation<T>(_ value: T) {
+    _print_unlocked(value, &self)
   }
+  
+  /// Creates a string from this instance, consuming the instance in the
+  /// process.
+  @inlinable
+  internal __consuming func make() -> String {
+    return _storage
+  }
+}
 
-  /// Creates a string containing the given value's textual representation.
-  ///
-  /// Do not call this initializer directly. It is used by the compiler when
-  /// interpreting string interpolations.
+extension DefaultStringInterpolation: CustomStringConvertible {
+  @inlinable
+  public var description: String {
+    return _storage
+  }
+}
+
+extension DefaultStringInterpolation: TextOutputStream {
   @inlinable
```

Comment:
...it might make sense to make this non-inlineable. If you'd like, I can make a throwaway branch to see how that benchmarks (against the current implementation).